src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -0,0 +1,175 @@
+            URI authorizationEndpoint = providerMetadata.getAuthorizationEndpointURI();
+            URI tokenEndpoint = providerMetadata.getTokenEndpointURI();
+
+            Scope supportedScopes = providerMetadata.getScopes();
+            List<String> scopesList = idpServerInfo.getRequestScopes();
+            Scope requestedScopes = new Scope();
+            if (scopesList != null) {
+                for (String scope : scopesList) {
+                    if (supportedScopes != null && supportedScopes.contains(scope)) {
+                        requestedScopes.add(new Scope.Value(scope));
+                    } else {
+                        logger.warning(
+                                "Requested scope '" + scope + "' is not supported by the IdP");
+                    }
+                }
+            }
+            // mongodb is not configured to ask for offline_access by default. We prefer always getting a
+            // refresh token when the server allows it.
+            if (!scopesList.contains(OFFLINE_ACCESS)) {
+                if (supportedScopes != null && supportedScopes.contains(OFFLINE_ACCESS)) {
+                    requestedScopes.add(new Scope.Value(OFFLINE_ACCESS));
+                } else {
+                    logger.info(
+                            "Offline access (refresh token) is not supported by the OIDC provider");
+                }
+            }
+
+            // Start the local RFC8252 HTTP server to receive the redirect.
+            server.start();
+
+            URI redirectURI =
+                    new URI(
+                            "http://localhost:"
+                                    + RFC8252HttpServer.DEFAULT_REDIRECT_PORT
+                                    + "/redirect");
+            State state = new State();
+            CodeVerifier codeVerifier = new CodeVerifier();
+
+            // Build the authorization request URI.
+            AuthorizationRequest request =
+                    new AuthorizationRequest.Builder(
+                                    new ResponseType(ResponseType.Value.CODE),
+                                    new ClientID(clientID))
+                            .scope(requestedScopes)
+                            .redirectionURI(redirectURI)
+                            .state(state)
+                            .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                            .endpointURI(authorizationEndpoint)
+                            .build();
+
+            // Open the browser to the authorization request URI.
+            if (Desktop.isDesktopSupported()) {
+                Desktop.getDesktop().browse(request.toURI());
+            } else {
+                logger.severe("Desktop operations not supported");
+                return null;
+            }
+
+            // Wait for the authorization response from the local HTTP server.
+            OidcResponse response = server.getOidcResponse(Duration.ofMinutes(5));
```

Comment:
Let's only use 5 minutes when the `OidcCallbackContext` does not contain a timeout.

Reply:
Thanks, I meant to do that after Natacha's comment.  It's set now.